### PR TITLE
Refactor Erl_type

### PR DESCRIPTION
**Pull request by youxkei**
- Introduce inline records
- Use specific data type instead of option

### CODE REVIEW

File: lib/erl_type.ml [hunk 1]
```diff
@@ -102,37 +105,38 @@ v}
...
     args : t list;
     struct_ : t;
   }
-[@@deriving show, sexp_of]
 and tuple = {
     types : t list;      (* type of elements *)
     arity : int;         (* size *)
     tag : string option; (* first element's tag (may be record name) *)
   }
+and tuple_or_any_tuple = TupleOrAnyTupleTuple of tuple | TupleOrAnyTupleAnyTuple
```
Comment (yoshihiro503) [inline lib/erl_type.ml]: How about `tuple_or_any_tuple = Tuple of tuple | AnyTuple` ?

File: lib/erl_type.ml [hunk 2]
```diff
@@ -26,13 +26,16 @@ let qualifier_of_etf etf =
...
 type number =
-  | IntRange of int option * int option (* (min or -infinity, max or infinity) *)
-  | IntSet of int list
+  | IntRange of {min: int_or_neg_inf; max: int_or_pos_inf}
+  | IntSet of {set: int list}
   | Integer (* integer() *)
   | Float   (* float()   *)
   | All     (* number()  *)
+and int_or_neg_inf = IntOrNegInfInt of int | IntOrNegInfNegInf
+and int_or_pos_inf = IntOrPosInfInt of int | IntOrPosInfPosInf
```
Comment (yoshihiro503) [inline lib/erl_type.ml]: How about

```ocaml
and min = Min of int | NegativeInfinity
and max = Max of int | Infinity
```

File: lib/erl_type.ml [hunk 3]
```diff
@@ -151,47 +155,55 @@ let rec of_etf = function
      | AtomTag ->
         begin match elements with
         | Etf.Atom "any" ->
-           Ok (Atom ["any"])
+            Ok (Atom {atoms_union=["any"]})
```
Comment (yoshihiro503) [inline lib/erl_type.ml]: Sorry this is a bug from before: this case is `atom()` not `'any'`.
Could you add a type represent `AtomsUnion` or `AnyAtom` and fix it?

File: lib/erl_type.ml (hunk 2 above)
Comment (youxkei) [reply lib/erl_type.ml]: Your naming looks much better than mine. I fixed to use these naming.

File: lib/erl_type.ml (hunk 3 above)
Comment (youxkei) [reply lib/erl_type.ml]: I fixed to handle `atom()` type correctly.

**Comment by youxkei** [comment]
@yoshihiro503 I came up with a better solution about AnyAtom and AnyTuple.
I defined AnyAtom and AnyTuple constructor in `Erl_type.t`.
With this change, we have no need to introduce redundant type such as `atoms_union_or_any_atom` and `tuple_or_any_tuple`.

What do you think about it?

**Comment by yoshihiro503** [comment]
It's so good. But for line https://github.com/dwango/fialyzer/pull/134/files#diff-23cc1503568b4380cc70164102261c93R121, it may be better just to do as follows simply.

```ocaml
  | Tuple of tuple
```

**Pull request merged**